### DISCUSSION TITLE
Power ON device, 5 second delay, power OFF if current is at or below threshold

### DISCUSSION THREAD

**Discussion by 49studebaker** [arendst/Tasmota#19062] [Q&A]
After the water pump is powered ON, wait 5 seconds before checking if the device is at or below the 0.20amp limit. After 5 seconds, if the water pump is at or below the limit, turn OFF the water pump.

Rule1
ON Energy#Current<=0.20 DO RuleTimer 5 ENDON
ON Energy#Current>0.20 DO RuleTimer 0 ENDON
ON Rules#Timer=1 DO Power1 0 ENDON

The console shows that Tasmota is executing "Energy#Current<=0.20 DO RuleTimer 5" while the device is turned OFF. Is there anyway to execute this code only when the device is turned ON?

**Comment by sfromis** [comment]
It may not really matter to have the rules running, doing no harm.

Anyway, one method could be to have the low-threshold in a var, which you can reset to 0 when powering off, and to 0.20 again when powering on.

**Reply by 49studebaker** [reply]
Rule1

ON Power1#State=0 DO Var2 -1 ENDON
ON Power1#State=1 DO Var2 0.20 ENDON

ON Energy#Current<=Var2 DO RuleTimer 5 ENDON
ON Energy#Current>Var2 DO RuleTimer 0 ENDON
ON Rules#Timer=1 DO Power1 0 ENDON


Why does it execute Ruletimer 5, when Var2 is set to -1 on power off. 0 <= -1 would be false, correct?

**Reply by barbudor** [reply]
It's may be a race condition which makes the trigger execute before the variable is actually set
But without detailed logs a level 3,I t's hard to guess

**Reply by 49studebaker** [reply]
> It's may be a race condition which makes the trigger execute before the variable is actually set
> 
> But without detailed logs a level 3,I t's hard to guess

How do I do a level 3?

**Reply by barbudor** [reply]
Weblog 3 for the web console

**Reply by barbudor** [reply]
Just wondering how is it a problem to have RuleTimer 5 in that case
It will issue a power off in 5 secs while it is already off. And so what?

But if you want to avoid that the RuleTimer triggers when the power is already off,  you may want to use a 'backlog ruletimer 0; var2 -1' just to be on the safe side.

**Reply by sfromis** [reply]
This will not work:
`ON Energy#Current<=Var2 DO RuleTimer 5 ENDON`
Should be:
`ON Energy#Current<=%Var2% DO RuleTimer 5 ENDON`

**Reply by 49studebaker** [reply]
> This will not work:
> 
> `ON Energy#Current<=Var2 DO RuleTimer 5 ENDON`
> 
> Should be:
> 
> `ON Energy#Current<=%Var2% DO RuleTimer 5 ENDON`

Could you explain why I need a % on each side of Var2?


Why doesn't the following line of code need a % on each side of Var2?

ON Energy#Current>Var2 DO RuleTimer 0 ENDON


Thanks.

**Reply by barbudor** [reply]
Actually both does
Var2 is a command to set the value in variable number 2
%var2% is the actual value of the variable

Good catch from @sfromis as usual 🔍

**Reply by sfromis** [reply]
Using `%` is to expand the value of the var before evaluating the rule, allowing a var to be used "anywhere" without the parser needing to be able to guess that `>Var2` does not mean comparing to the raw string `Var2`.

To illustrate the point, this silly example would work if first having done `var5 er 0`
`ON Energy#Current>%Var2% DO RuleTim%var5% ENDON`
Using vars is not really part of the syntax of the rules language, except when building with the optional feature of expressions.
And then you have a number of commands specifically to modify var values, like `var`, `add`, `sub`, `mult` and `scale`.

**Reply by 49studebaker** [reply]
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F4, Count 552
00:00:00.055 SER: Set to 8E1 4800 bit/s
00:00:00.179 QPC: Count 1
00:00:00.196 NRG: Init driver 2
00:00:00.213 Project tasmota - Tasmota Version 12.5.0(tasmota)-2_7_4_9(2023-04-17T08:03:54)
00:00:00.218 SER: Set to 8E1 4800 bit/s
00:00:00.221 SNS: Hardware Serial
00:00:00.776 WIF: Connecting to AP1 CH iPhone Channel 6 BSSId AA:75:CD:F7:E0:25 in mode 11n as tasmota-6632D7-4823...
00:00:01.550 WIF: Connected
00:00:01.763 HTP: Web server active on tasmota-6632D7-4823 with IP address 172.20.10.2
19:23:19.413 RSL: INFO1 = {"Info1":{"Module":"Sonoff S31","Version":"12.5.0(tasmota)","FallbackTopic":"cmnd/DVES_6632D7_fb/","GroupTopic":"cmnd/tasmotas/"}}
19:23:19.429 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-6632D7-4823","IPAddress":"172.20.10.2"}}
19:23:19.446 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":75}}
19:23:19.462 RSL: RESULT = {"POWER":"OFF"}
19:23:19.463 RSL: POWER = OFF
19:23:19.355 RUL: ENERGY#CURRENT<=%VAR2% performs "RuleTimer 5"
19:23:19.361 RSL: RESULT = {"T1":5,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
19:23:22.226 QPC: Reset
19:23:23.135 RSL: STATE = {"Time":"2023-07-08T19:23:23","Uptime":"0T00:00:08","UptimeSec":8,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"CH iPhone","BSSId":"AA:75:CD:F7:E0:25","Channel":6,"Mode":"11n","RSSI":80,"Signal":-60,"LinkCount":1,"Downtime":"0T00:00:03"}}
19:23:23.145 RSL: SENSOR = {"Time":"2023-07-08T19:23:23","ENERGY":{"TotalStartTime":"2023-05-25T22:26:44","Total":0.000,"Yesterday":0.000,"Today":0.000,"Period":0,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000}}
19:23:25.154 RUL: RULES#TIMER=1 performs "Power1 0"
19:23:25.159 RSL: RESULT = {"POWER":"OFF"}
19:23:25.161 RSL: POWER = OFF

SmartPlug turned ON. Current remains 0 until PulseTime 32 second shutoff. SmartPlug turns OFF.

19:25:23.034 RSL: RESULT = {"POWER":"ON"}
19:25:23.036 RSL: POWER = ON
19:25:23.080 RUL: POWER1#STATE=1 performs "Var2 0.20"
19:25:23.085 RSL: RESULT = {"Var2":"0.20"}
19:25:55.102 RSL: RESULT = {"POWER":"OFF"}
19:25:55.104 RSL: POWER = OFF
19:25:55.206 RUL: POWER1#STATE=0 performs "Var2 -1"
19:25:55.212 RSL: RESULT = {"Var2":"-1"}
19:25:55.249 RUL: ENERGY#CURRENT>%VAR2% performs "RuleTimer 0"
19:25:55.255 RSL: RESULT = {"T1":0,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}

**Reply by sfromis** [reply]
Well, I was only suggesting the **low** threshold to be using the var. Same var for the high threshold means that it is always higher than the threshold if you set it it to -1.

**Reply by 49studebaker** [reply]
Rule2 “once” is OFF

Rule2
ON System#Init DO Var3 0 ENDON

ON Power1#State=1 DO Var3 0.20 ENDON

ON Energy#Current<%Var3% DO Backlog RuleTimer 5; Var3 0 ENDON

ON Energy#Current>0.20 DO Backlog RuleTimer 0; Var3 0.20 ENDON

ON Rules#Timer=1 DO Power1 0 ENDON
